src/nrfx/hal/nrf_egu.c

```diff
@@ -0,0 +1,176 @@
+/** Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Note that the function prototypes are taken from the NRFx HAL
+ */
+
+ #if 1
+
+#include "hal/nrf_egu.h"
+#include "bs_tracing.h"
+#include "NRF_EGU.h"
+#include <kernel.h>
+#include <irq_ctrl.h>
+
+uint32_t nrf_egu_channel_count(NRF_EGU_Type const *p_reg)
+{
+	return 16;
+}
+
+static int reg_to_irq(NRF_EGU_Type *p_reg)
+{
+	extern NRF_EGU_Type NRF_EGU_regs[6];
+	size_t raw_offset = (size_t)p_reg - (size_t)(&NRF_EGU_regs[0]);
+	size_t id = raw_offset / sizeof(NRF_EGU_Type);
+
+	switch (id) {
+	case 0: return SWI0_EGU0_IRQn; break;
+	case 1: return SWI1_EGU1_IRQn; break;
+	case 2: return SWI2_EGU2_IRQn; break;
+	case 3: return SWI3_EGU3_IRQn; break;
+	case 4: return SWI4_EGU4_IRQn; break;
+	case 5: return SWI5_EGU5_IRQn; break;
+	}
+	return 0xff;
+}
+
+static uint32_t task_to_id(nrf_egu_task_t task)
+{
+	return task / sizeof(((NRF_EGU_Type *)NULL)->TASKS_TRIGGER[0]);
+}
+
+void nrf_egu_task_trigger(NRF_EGU_Type *p_reg, nrf_egu_task_t egu_task)
+{
+	p_reg->EVENTS_TRIGGERED[task_to_id(egu_task)] = 1;
```

Comment:
I guess you also need to check if the event is connected to a ppi channel, and if it is, trigger the corresponding task?

Reply:
I do not understand this part, on real hardware when I write to TASK_TRIGGER register, it will trigger an EGU IRQ. In handler of this IRQ I check the state of EVENTS_TRIGGERED, and event corresponding to the task is selected. So if I understand correctly, I write to the TASK_TRIGGER, and it results in immediate write to EVENTS_TRIGGERED and calling EGU IRQ if the interrupt for this task is enabled. 

In the simulated EGU I just ignored the TASK_TRIGGER register, because I did not see any use for it. 

Do You have a test case in mind that would work on nrf52840, and not on bableSim with my implementation?

Reply:
Maybe you could try the following:
```
NRF_PPI->CH[0].EEP = (uint32_t) NRF_EGU0->EVENTS_TRIGGERED[0];
NRF_PPI->CH[0].TEP = (uint32_t) NRF_TIMER0->TASKS_START;

NRF_EGU0->TASK_TRIGGER[0] = 1;
```
I didn't check it myself, but I would expect the TIMER to be started

Reply:
Hmm, I don't really know how PPI works, but from your example I see that It works like a pointer, first line registering a watch on variable, and another is a register that has to follow the state of watched register ? 

So when EVENTS_TRIGGERED is changed by EGU, it changed EEP in PPI, and it is passed to TEP of PPI, and because timer register is connected there, it starts the timer. Do I understand it correctly ? 

I definitely need to read more about PPI, because Your snippet looks like magic to me :D

Reply:
Hmm, PPI as far as I understand provides a service to mirror value from one register to another, so when the EVENTS_TRIGGERED[0] register is changed, it triggers a change in TASKS_START of timer. 

This means, that whenever I modify EVENTS_TRIGGERED, I also need to modify TASKS_TRIGGER, because it can be connected to PPI.

I can see a major issue with PPI, because it can work in another direction too, a timer can trigger an event, and EGU may be connected to TEP register. In this example PPI would directly modify the EGU register, and in simulation this is just a variable, so it will not trigger the same action as using the function to trigger event

I think that to correctly simulate EGU, I need to add a function that would be called periodically, and it will enable/ disable IRQ based on the values in EGU register, This function would have to be called in main loop of the simulation. 

Did I understand correctly Your concerns ?

Reply:
I wonder if this would be a good approach for any core peripheral ? Just use default implementations of HAL functions, but add a routine to check the register state periodically, and therefore simulate the peripheral based on the state of registers?

The perfect solution would be to attach a trigger function for every memory modification, like a data break point in debugger, but I do not know if something like this exists in plain C on native posix ?

Reply:
@RobertGalatNordic 
In general register sideeffects are modeled as described here:
https://github.com/BabbleSim/ext_NRF52_hw_models/blob/master/docs/README_HW_models.md#the-sw-registers-if
(You don't want to periodically poll). You can see the models include some nRFx HAL function replacements that have this special sideeffects calls.
You can take a look at the PPI model to see how the tasks and events are plugged ( nrf_ppi_event() is called from other HW models when they raise an event)

Reply:
@aescolar  thank You for the explanation, I still have one concern. In the application the user can directly modify the registers of a peripheral, in the bsim it is just a variable. I understand that PPI will trigger sideffects for target peripheral, but it will not happen if in application code register is modified directly. How can we handle this case ?

Reply:
We just don't support it :-) : it is not supported that an app modifies a register with sideeffects directly. The app should use the HAL, or if they really really want to access a register with sideeffects directly they can add a ifdef'ed call to the HW model sideeffect function.